lib/OpenStackExtract/MiqOpenStackVm/MiqOpenStackInstance.rb

```diff
@@ -0,0 +1,188 @@
+    log_prefix = "MIQ(#{self.class.name}##{__method__}) instance_id=[#{@instance_id}]"
+
+    miq_snapshot = nil
+    if snapshot_image_id
+      $log.debug "#{log_prefix}: Found pointer to existing snapshot: #{snapshot_image_id}"
+      miq_snapshot = begin
+        image_service.images.get(snapshot_image_id)
+      rescue => err
+        $log.debug "#{log_prefix}: #{err}"
+        $log.debug err.backtrace.join("\n")
+        nil
+      end
+
+      if miq_snapshot
+        raise "Already has an EVM snapshot: #{miq_snapshot.name}"
+      else
+        $log.debug "#{log_prefix}: Snapshot does not exist, deleting metadata"
+        snapshot_metadata.destroy
+      end
+    else
+      $log.debug "#{log_prefix}: No existing snapshot detected for: #{instance.name}"
+    end
+
+    $log.debug "#{log_prefix}: Snapshotting instance: #{instance.name}..."
+    # TODO: pass in snapshot name.
+    rv = compute_service.create_image(instance.id, "EvmSnapshot", :description => options[:desc])
+    rv.body['image'][:service] = image_service
+    miq_snapshot = Fog::Image::OpenStack::Image.new(rv.body['image'])
+    miq_snapshot.collection = image_service.images
+
+    until miq_snapshot.status == "active"
+      $log.debug "#{log_prefix}: #{miq_snapshot.status}"
+      sleep 1
+      miq_snapshot.reload
+    end
+    $log.debug "#{log_prefix}: #{miq_snapshot.status}"
+    $log.debug "#{log_prefix}: EVM snapshot creation complete"
+
+    instance.metadata.update(:miq_snapshot => miq_snapshot.id)
+    return miq_snapshot
+  rescue => err
+    $log.error "#{log_prefix}, error: #{err}"
+    $log.debug err.backtrace.join("\n") if $log.debug?
+    raise
+  end
+
+  def delete_evm_snapshot(image_id)
+    log_prefix = "MIQ(#{self.class.name}##{__method__}) snapshot=[#{image_id}]"
+
+    snapshot = begin
+      image_service.images.get(image_id)
+    rescue => err
+      $log.debug "#{log_prefix}: #{err}"
+      $log.debug err.backtrace.join("\n")
+      nil
+    end
+
+    if snapshot
+      begin
+        if snapshot_image_id != image_id
+          $log.warn "#{log_prefix}: pointer from instance doesn't match #{snapshot_image_id}"
+        end
+        $log.info "#{log_prefix}: deleting snapshot image"
+        snapshot.destroy
+        snapshot_metadata.destroy
+      rescue => err
+        $log.debug "#{log_prefix}: #{err}"
+        $log.debug err.backtrace.join("\n") if $log.debug?
+      end
+    else
+      $log.info "#{log_prefix}: no longer exists, deleting references"
+    end
+  end
+
+  private
+
+  def miq_vm
+    raise "Instance: #{instance.id}, does not have snapshot reference." unless snapshot_image_id
+    @miq_vm ||= begin
+      @temp_image_file = get_image_file(snapshot_image_id)
+      hardware  = "scsi0:0.present = \"TRUE\"\n"
+      hardware += "scsi0:0.filename = \"#{@temp_image_file.path}\"\n"
+      MiqVm.new(hardware)
+    end
+  end
+
+  def get_image_file(image_id)
+    log_pref = "#{self.class.name}##{__method__}"
```

Comment:
Most of the code in lib doesn't adhere to that pattern. As you mentioned before, logging in lib needs to be addressed in a general way as we carve the code into gems.